### ISSUE TITLE
Error connecting to external APIs from a Micronaut application deployed to AWS Lambda using custom runtime

### ISSUE THREAD

**Issue by ilopmar** [oracle/graal#1858]
We have found an issue running a Micronaut application as a GraalVM native image deployed in AWS Lambda using a custom runtime.
Everything works properly, but if the application executes a call to an external API, there is a connection time out.
We've been trying and testing different things and also different external APIs and it doesn't work.

Things that work: Not using Micronaut HTTP Client to send the request, deploy the application as JVM to AWS Lambda and use Java8 runtime, "convert" the application to a normal Micronaut + GraalVM application.

We've contacted the AWS team and Stefano Buliani (@sapessi) has been helping us to figure it out what's going on. He thinks there is some issue with Netty and GraalVM.

We have this issue in Micronaut: https://github.com/micronaut-projects/micronaut-core/issues/2335 in which you can see some of Stefano's comments.

Also, he updated us last night with the following:

> One more update from our side. We created the new test environment for the “provided” runtime and strace-ed both the jar running in the stock JVM and the graalvm-compiled executable. So far as I can tell, the primary difference is that the native executable gives up at the first attempt to the futex(FUTEX_WAIT) call for the epoll even though it returns EAGAIN whereas the JVM retries the call and it succeeds. 

```
Native executable:
[pid    20] connect(102, {sa_family=AF_INET, sin_port=htons(80), sin_addr=inet_addr(\"35.170.171.200\")}, 16) = -1 EINPROGRESS (Operation now in progress)\n",
[pid    20] epoll_ctl(57, EPOLL_CTL_ADD, 102, {EPOLLOUT, {u32=102, u64=15912961789977952358}}) = 0\n",
[pid    20] epoll_ctl(57, EPOLL_CTL_MOD, 102, {EPOLLOUT, {u32=102, u64=4294967398}}) = 0\n",
[pid    20] epoll_wait(57, {{EPOLLIN, {u32=55, u64=15912961377661091895}}}, 1024, 0) = 1\n",
pid    20] futex(0x41ea784, FUTEX_CMP_REQUEUE_PRIVATE, 1, 2147483647, 0x41ea750, 4 )       = 1\n"
[pid    20] futex(0x41ea750, FUTEX_WAKE_PRIVATE, 1 )       = 0\n
[pid    20] getsockopt(102, SOL_SOCKET, SO_LINGER,  {onoff=0, linger=0}, [8]) = 0\n",
# pid 13 forked pid 20 earlier on
[pid    13] futex(0x41ea750, FUTEX_WAIT_PRIVATE, 2, NULL  )       = -1 EAGAIN (Resource temporarily unavailable)\n",
# At this point pid 13 begins dumping the exception (write())

 
JVM version
32    socket(PF_INET, SOCK_DGRAM|SOCK_CLOEXEC|SOCK_NONBLOCK, IPPROTO_IP) = 43
32    connect(42, {sa_family=AF_INET, sin_port=htons(80), sin_addr=inet_addr("52.7.202.148")}, 16) = -1 EINPROGRESS (Operation now in progress)
32    futex(0x7fbd60158c54, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x7fbd60158c50, {FUTEX_OP_SET, 0, FUTEX_OP_CMP_GT, 1}) = 1
32    futex(0x7fbd6041e028, FUTEX_WAKE_PRIVATE, 1)             = 0
32    epoll_ctl(32, EPOLL_CTL_ADD, 42, {EPOLLOUT, {u32=42, u64=7399571624090927146}}) = 0
32    epoll_ctl(32, EPOLL_CTL_MOD, 42, {EPOLLOUT, {u32=42, u64=42}}) = 0
32    epoll_wait(32, {{EPOLLIN, {u32=30, u64=52776558133278}}}, 1024, 29990) = 1
32    read(30, "\1", 128)               = 1
32    epoll_wait(32,  <unfinished ...>
32    getsockopt(42, SOL_SOCKET, SO_ERROR, [0], [4]) = 0
32    getsockname(42, {sa_family=AF_INET, sin_port=htons(48748), sin_addr=inet_addr("169.254.76.1")}, [16]) = 0
32    getsockname(42, {sa_family=AF_INET, sin_port=htons(48748), sin_addr=inet_addr("169.254.76.1")}, [16]) = 0
32    futex(0x7fbd6019ff54, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x7fbd6019ff50, {FUTEX_OP_SET, 0, FUTEX_OP_CMP_GT, 1}) = 1
32    futex(0x7fbd6019ff28, FUTEX_WAKE_PRIVATE, 1 )             = 0
22    futex(0x7fbd6019ff28, FUTEX_WAIT_PRIVATE, 2, NULL )             = -1 EAGAIN (Resource temporarily unavailable)
22    futex(0x7fbd6019ff28, FUTEX_WAKE_PRIVATE, 1) = 0
```

> Correction, the native executable just seems not to handle the EAGAIN gracefully. Apologies for the mistake.

In order to reproduce the issue clone the following repo: https://github.com/ilopmar/chuck-norris-facts-mn-graal-function

The `master` branch contains the example.
If you run `deploy.sh` the application is built using docker and deployed to AWS lambda using Cloud Formation. Please take a look at the script because you will need to define your own S3 bucket. In this branch we use GraalVM 19.2.1.
After it is deployed you will see the url in which the lambda is exposed so just send a curl: `curl https://5rjbh51agf.execute-api.eu-west-1.amazonaws.com/Prod/joke`. You will get almost instantaneously a `{"message":"Gateway timeout"}` and in Cloudwatch you'll see the timeout (check the Micronaut issue I linked before to see the logs)

I have also updated to GraalVM 19.3.0 in branch `19.3.0`. I've tried with and without including libsunec.so (just in case) but the error is the same, timeout.

Hopefully you can help us an Stefano to figure out what's going on.

Thanks!

**Comment by sapessi** [comment]
@Petermd also correctly pointed out that the `epoll_wait` call in the native executable does not set a timeout, whereas the JVM on does.

**Comment by christianwimmer** [comment]
You should certainly drop all work and  analysis on GraalVM 19.2 and only use GraalVM 19.3. The 19.3 version now uses the C code of the JDK, so if there were any bugs in the manual substitutions uses in the 19.2 version that is no longer relevant.

But it seems like you observe the problem also on 19.3

**Comment by petermd** [comment]
Re-verified with GraalVM 19.3 and still see the same behavior with immediate timeout. Will dig deeper

```
19    00:36:43.868112 connect(102, {sa_family=AF_INET, sin_port=htons(80), sin_addr=inet_addr("3.221.56.183")}, 16) = -1 EINPROGRESS (Operation now in progress)                  
// connect is async, add fd to epoll 
19    00:36:43.868235 epoll_ctl(56, EPOLL_CTL_ADD, 102, {EPOLLOUT, {u32=102, u64=33636964091560038}}) = 0                                                                         
19    00:36:43.868321 epoll_ctl(56, EPOLL_CTL_MOD, 102, {EPOLLOUT, {u32=102, u64=33636964091560038}}) = 0                                                                         
19    00:36:43.868395 epoll_wait(56, {{EPOLLIN, {u32=54, u64=33636964091559990}}}, 4096, 0) = 1                                                                                   
// epoll_wait has deadline=0, but only 1 fd (54, not 102) has event, so we read
19    00:36:43.868470 read(54, "\1", 128) = 1                                                                                                                                     
// no follow on epoll_wait, suggests we are executing the connect failure Runnable now
19    00:36:43.868624 futex(0x39a2324, FUTEX_CMP_REQUEUE_PRIVATE, 1, 2147483647, 0x39a22f0, 4 <unfinished ...>                                                                    
13    00:36:43.868672 <... futex resumed> ) = 0                                                                                                                                   
19    00:36:43.868686 <... futex resumed> ) = 1                                                                                                                                   
13    00:36:43.868700 futex(0x39a22f0, FUTEX_WAIT_PRIVATE, 2, NULL <unfinished ...>                                                                                               
19    00:36:43.868715 futex(0x39a22f0, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>                                                                                                     
13    00:36:43.868730 <... futex resumed> ) = -1 EAGAIN (Resource temporarily unavailable)                                                                                        
19    00:36:43.868746 <... futex resumed> ) = 0                                                                                                                                   
13    00:36:43.868760 futex(0x39a22f0, FUTEX_WAKE_PRIVATE, 1) = 0                                                                                                                 
13    00:36:43.869071 mmap(NULL, 2097152, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0 <unfinished ...>                                                                 
19    00:36:43.869099 getsockopt(102, SOL_SOCKET, SO_LINGER,  <unfinished ...>                                                                                                    
13    00:36:43.869116 <... mmap resumed> ) = 0x7f3b2fbe7000                                                                                                                       
19    00:36:43.869132 <... getsockopt resumed> {onoff=0, linger=0}, [8]) = 0                                                                                                      
13    00:36:43.869152 munmap(0x7f3b2fbe7000, 102400) = 0                                                                                                                          
13    00:36:43.869218 munmap(0x7f3b2fd00000, 946176) = 0                                                                                                                          
// start to write exception
13    00:36:43.869391 write(2, "io.micronaut.http.client.excepti"..., 126 <unfinished ...>                                                                                        
```

**Comment by ilopmar** [comment]
Finally @petermd found the issue in Netty configuration for GraalVM, everything is explained here https://github.com/micronaut-projects/micronaut-core/issues/2335#issuecomment-570151944.  
We've released a new version of Micronaut (1.3.0.RC1) with the fix on our side and the fix has also been merged into Netty (https://github.com/netty/netty/pull/9963) and it will be available in next release (4.1.46).

Thank you everyone for your help :smiley:
